code/modules/shuttle/custom_shuttle.dm

```diff
@@ -0,0 +1,229 @@
+		dat += "Sublight Speed: [calculated_speed]ms<sup>-1</sup><br>"
+		dat += calculated_speed < 1 ? "<b>INSUFFICIENT ENGINE POWER</b><br>" : ""
+		dat += calculated_non_operational_thrusters > 0 ? "<b>Warning: [calculated_non_operational_thrusters] thrusters offline.</b><br>" : ""
+		dat += "Fuel Consumption: [calculated_consumption]units per distance<br>"
+		dat += "Engine Cooldown: [calculated_cooldown]s<hr>"
+		var/destination_found
+		for(var/obj/docking_port/stationary/S in SSshuttle.stationary)
+			if(!options.Find(S.id))
+				continue
+			if(!M.check_dock(S, silent=TRUE))
+				continue
+			if(calculated_speed == 0)
+				break
+			destination_found = TRUE
+			var/dist = round(calculateDistance(S))
+			dat += "<A href='?src=[REF(src)];setloc=[S.id]'>Target [S.name] (Dist: [dist] | Fuel Cost: [round(dist * calculated_consumption)] | Time: [round(dist / calculated_speed)])</A><br>"
+		if(!destination_found)
+			dat += "<B>No valid destinations</B><br>"
+		dat += "<hr>[targetLocation ? "Target Location : [targetLocation]" : "No Target Location"]"
+		dat += "<hr><A href='?src=[REF(src)];fly=1'>Initate Flight</A><br>"
+	dat += "<A href='?src=[REF(user)];mach_close=computer'>Close</a>"
+
+	popup = new(user, "computer", M ? M.name : "shuttle", 350, 450)
+	popup.set_content("<center>[dat]</center>")
+	popup.set_title_image(usr.browse_rsc_icon(src.icon, src.icon_state))
+	popup.open()
+
+/obj/machinery/computer/custom_shuttle/Topic(href, href_list)
+	if(..())
+		return
+	usr.set_machine(src)
+	src.add_fingerprint(usr)
+	if(!allowed(usr))
+		to_chat(usr, "<span class='danger'>Access denied.</span>")
+		return
+
+	if(href_list["calculate"])
+		calculateStats()
+		ui_interact(usr)
+		return
+	var/obj/docking_port/mobile/M = SSshuttle.getShuttle(shuttleId)
+	if(!M)
+		return
+	if(M.launch_status == ENDGAME_LAUNCHED)
+		return
+	if(href_list["setloc"])
+		SetTargetLocation(href_list["setloc"])
+		ui_interact(usr)
+		return
+	else if(href_list["fly"])
+		Fly()
+		ui_interact(usr)
+		return
+
+/obj/machinery/computer/custom_shuttle/proc/calculateDistance(var/obj/docking_port/stationary/port)
+	var/deltaX = port.x - x
+	var/deltaY = port.y - y
+	var/deltaZ = (port.z - z) * Z_DIST
+	return sqrt(deltaX * deltaX + deltaY * deltaY + deltaZ * deltaZ)
+
+/obj/machinery/computer/custom_shuttle/proc/linkShuttle(var/new_id)
+	shuttleId = new_id
+	possible_destinations = "shuttle[new_id]_custom"
+
+/obj/machinery/computer/custom_shuttle/proc/calculateStats(var/useFuel = FALSE, var/dist = 0)
+	var/obj/docking_port/mobile/M = SSshuttle.getShuttle(shuttleId)
+	if(!M)
+		return FALSE
+	//Reset data
+	calculated_mass = 0
+	calculated_dforce = 0
+	calculated_speed = 0
+	calculated_engine_count = 0
+	calculated_consumption = 0
+	calculated_cooldown = 0
+	calculated_fuel_less_thrusters = 0
+	calculated_non_operational_thrusters = 0
+	//Calculate all the data
+	var/list/areas = M.shuttle_areas
+	for(var/shuttleArea in areas)
+		for(var/each in shuttleArea)
```

Comment:
I know there is a better method for this, will have to codedive to find something else

Reply:
Could just fudge it a bit and go by the number of turfs in the area.

Reply:
But if you want to make it slightly more complex you could consider closed turfs (walls, etc.) as having more mass than open turfs (floors, etc.)

Reply:
yea currently every single object and turf weighs 100kg

Reply:
I want machines to weight, so I have it so turfs, machines and structures weigh 100kg (about average for a shuttle's mass to make sense)

Reply:
I'm still not a fan of how expensive looping over literally everything is.

Reply:
Alright I'll figure something out, have got a spam prevention thing on it

Reply:
Alright, got a nifty patch up.
Instead of looping through every atom on the turf, we get the amount of atoms on the turf and use that for the weight, only looping through every engine on the turf.